src/contexts/rag-context.tsx

```diff
@@ -19,40 +18,28 @@ export type RAGContextType = {
 const RAGContext = createContext<RAGContextType | null>(null)
 
 export function RAGProvider({
-  setPluginRAGEngine,
   children,
-}: PropsWithChildren<{ setPluginRAGEngine: (ragEngine: RAGEngine) => void }>) {
+  onRAGEngineChange,
+}: PropsWithChildren<{ onRAGEngineChange: (ragEngine: RAGEngine) => void }>) {
   const app = useApp()
   const { settings } = useSettings()
+  const [ragEngine, setRagEngine] = useState<RAGEngine | null>(null)
 
-  const ragEngine = useMemo(() => {
-    return new RAGEngine(app, settings)
+  useEffect(() => {
+    RAGEngine.create(app, settings).then(setRagEngine)
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [app])
 
-  const embeddingModel = useMemo(() => {
-    return getEmbeddingModel(settings.embeddingModel, {
-      openAIApiKey: settings.openAIApiKey,
-    })
-  }, [settings.embeddingModel, settings.openAIApiKey])
-
   useEffect(() => {
-    void ragEngine.initialize(embeddingModel)
-    setPluginRAGEngine(ragEngine)
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [ragEngine])
+    if (ragEngine) {
+      onRAGEngineChange(ragEngine)
+    }
+  }, [ragEngine, onRAGEngineChange])
 
   useEffect(() => {
     ragEngine?.setSettings(settings)
   }, [ragEngine, settings])
```

Comment:
Do not create RagEngine again in here. Just have it passed from the Plugin where it was created. And when settings change, there's no need to call ragEngine.setSettings here; it can be handled in Plugin.setSettings.